src/components/views/login/ServerConfig.js

```diff
@@ -46,19 +59,21 @@ module.exports = React.createClass({
 
     getInitialState: function() {
         return {
-            hs_url: this.props.defaultHsUrl,
-            is_url: this.props.defaultIsUrl,
-            original_hs_url: this.props.defaultHsUrl,
-            original_is_url: this.props.defaultIsUrl,
-            // no toggle button = show, toggle button = hide
-            configVisible: !this.props.withToggleButton
+            hs_url: this.props.customHsUrl,
+            is_url: this.props.customIsUrl,
+            // if withToggleButton is false, then show the config all the time given we have no way otherwise of making it visible
+            configVisible: !this.props.withToggleButton || 
+                           (this.props.customHsUrl !== this.props.defaultHsUrl) ||
+                           (this.props.customIsUrl !== this.props.defaultIsUrl)
         }
     },
 
     onHomeserverChanged: function(ev) {
         this.setState({hs_url: ev.target.value}, function() {
             this._hsTimeoutId = this._waitThenInvoke(this._hsTimeoutId, function() {
-                this.props.onHsUrlChanged(this.state.hs_url.replace(/\/$/, ""));
+                var hsUrl = this.state.hs_url.trim().replace(/\/$/, "");
```

Comment:
If `customHsUrl` is `null`, then `this.state.hs_url` will be `null`, meaning this will NPE. We should have default props for the custom URLs, or mark `isRequired` so when the inevitable happens we have an idea why.